src/Parser.php

```diff
@@ -12,6 +12,8 @@ interface Parser
     /**
      * Parses the JWT and returns a token
      *
+     * @param non-empty-string $jwt
```

Comment:
I think a parser should accept an empty string here, with `CannotDecodeContent` as an outcome :thinking:

Reply:
Need to think about it, but I see a `CannotDecodeContent` exception as quite valuable on an empty string scenario here: forcing further verification on the client side seems wasteful, at this stage.

I agree that a `string&empty` will always result in a crash here: just wondering whether forcing an additional upfront assertion is added value.

In the following (valid) code:

```php
function foo(string $token) {
    Assert::notEmpty($token);
    $parsed = $this->parser->parse($token);

    // ...
```

If a `CannotDecodeContent` is thrown (explaining that the content was empty) is the same value of `AssertionFailed`, except with an additional piece of code for it.

I'm just trying to figure out the ergonomics of the library here:

 * correctness leads to `string&empty -> never`
 * usability leads to `non-empty-string|(string&empty) -> Token|Exception`

Reply:
Now I understand better your doubts.

I can only say the sooner the type is safe, the better:

* if the library allows an empty string, my whole app may allow it
* if the library denies an empty string, I have to design it upfront to get rid of it. This basically means input validation before consumption

Reply:
There is one good thing coming from this new type restriction: the existence of the token must be verified by the consumer.

Anyway, I'm just presenting the options: I think that the parser (being a parser) is responsible for upcasting what comes in (or throwing), so it's our decision to decide how wide the API is supposed to be.

The parser makes the type safe anyway, so I have doubts on the value of an early assertion, in this specific case.

Reply:
That said, we can experiment with this, and then later-on relax the type constraint, if we consider it excessive.